### PR TITLE
Update Thanos to latest main

### PR DESCRIPTION
**Pull request by siggy**
Update Thanos dependency to include thanos-io/thanos#4928, to conserve memory.

Signed-off-by: Andrew Seigner <andrew@sig.gy>

**What this PR does**:

Bumps Thanos dependency to the latest main: https://github.com/thanos-io/thanos/commit/d1acaea2a11a3e4db6bb435c98dea63c517e3530

Most notably this pulls in thanos-io/thanos#4928, which gives a 2x~4x memory reduction when using Azure blocks storage:

cortex thanos bump

This PR, combined with #4581, brings Cortex's memory usage back inline with `v1.10.0`.

Relates to:
- #4581
- thanos-io/thanos#4928

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by bboreham** [comment]
Thanks for doing this.  It's best to cover all important changes when updating a dependency, so I checked
https://github.com/thanos-io/thanos/compare/de0e3848ff60...d1acaea2a11a
and I see that Redis is now a cache option. This sounds quite interesting; is it usable from Cortex now or does it need plumbing?

Please can you check the list and summarise all user-visible changes in the CHANGELOG.  Just in a few words - it's not necessary to link to individual PRs besides this one where a keen reader can find all the details.  Readers of CHANGELOG are assumed to be admins who don't want code-level details.

**Comment by siggy** [comment]
> I see that Redis is now a cache option. This sounds quite interesting; is it usable from Cortex now or does it need plumbing?

I think this would need plumbing. Cache instantation in Cortex:
https://github.com/cortexproject/cortex/blob/4474da8f28aefd5254d1d193df78e74c3ff9bcf4/pkg/storage/tsdb/caching_bucket.go#L152-L169
https://github.com/cortexproject/cortex/blob/4474da8f28aefd5254d1d193df78e74c3ff9bcf4/pkg/storage/tsdb/index_cache.go#L79-L89
... would need to call `cacheutil.NewRedisClient` in Thanos:
https://github.com/thanos-io/thanos/blob/9a26b0edee19a06c6e99a09e33ebceca734c91f9/pkg/cacheutil/redis_client.go#L122-L133

Interestingly the Thanos PR mentions Cortex's redis support as a reference implementation: thanos-io/thanos#4888

File: go.mod [hunk 1]
```diff
@@ -86,8 +83,8 @@ replace google.golang.org/grpc => google.golang.org/grpc v1.38.0
 // We only pin this version to avoid problems with running go get: github.com/thanos-io/thanos@main. That
 // currently fails because Thanos isn't merging release branches to main branch, and Go modules system is then
 // confused about which version is the latest one. v0.22.0 was released in July, but latest tag reachable from main
-// is v0.19.1. We pin version from late september here. Feel free to remove when updating to later version.
-replace github.com/thanos-io/thanos v0.22.0 => github.com/thanos-io/thanos v0.19.1-0.20211122085937-de0e3848ff60
+// is v0.19.1. We pin version from early December here. Feel free to remove when updating to later version.
```
[the remark this answers is not in the document]
Comment (siggy) [reply go.mod:86]: Thanos release tags are still not on the `main` branch. This `replace` pattern allowed me to simply put `main` on this line, and then `go mod vendor` correctly resolved it to `v0.19.1-0.20211122085937-de0e3848ff60`.

[the remark this answers is not in the document]
Comment (bboreham) [reply go.mod:86]: Thanks.
The comment above says it's about merging release branches, but this kinda looks like a merge of the release branch: https://github.com/thanos-io/thanos/pull/4738
Did it not work?  Not what Go is expecting?

**Review by bboreham — APPROVED** [review]
Thanks; LGTM

File: go.mod (hunk 1 above)
[the remark this answers is not in the document]
Comment (pstibrany) [reply go.mod:86]: That was a squash merge, not merge commit. It didn't make tag `v0.23.1` reachable from `main` branch.

[the remark this answers is not in the document]
Comment (pstibrany) [reply go.mod:86]: I tried to update release process instructions in Thanos at https://github.com/thanos-io/thanos/pull/4770 to hopefully fix this situation in the future.

[the remark this answers is not in the document]
Comment (siggy) [reply go.mod:86]: Thanks for filing thanos-io/thanos#4770 @pstibrany!

> Did it not work? Not what Go is expecting?

So we can do a `go get` on a release tag, and it works:
```bash
$ go get github.com/thanos-io/thanos@v0.24.0-rc.1
$ go mod vendor
$ git diff
...
-       github.com/thanos-io/thanos v0.22.0
+       github.com/thanos-io/thanos v0.24.0-rc.1
-replace github.com/thanos-io/thanos v0.22.0 => github.com/thanos-io/thanos v0.19.1-0.20211208205607-d1acaea2a11a
+// replace github.com/thanos-io/thanos v0.22.0 => github.com/thanos-io/thanos v0.19.1-0.20211208205607-d1acaea2a11a
```
... I don't know how important it is that we be able to do `go get github.com/thanos-io/thanos@main`, but that still doesn't work:
```bash
$ go get github.com/thanos-io/thanos@main
go: downloading github.com/thanos-io/thanos v0.19.1-0.20211213141611-c7a44e23febe
go get: github.com/thanos-io/thanos@main (v0.19.1-0.20211213141611-c7a44e23febe) requires github.com/thanos-io/thanos@v0.22.0, not github.com/thanos-io/thanos@main (v0.19.1-0.20211213141611-c7a44e23febe)
```

So I guess we have a few options:
1) Leave as-is, wait for thanos-io/thanos#4770.
2) Remove the `replace` and the comment, `require` `v0.24.0-rc.1` directly.
3) Remove the `replace` and the comment, wait for `v0.24.0` and `require` it directly.

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

**Pull request merged**